File: Packages/js-common/Source/DB/@Shared/TablePermissions.ts
import {CreateAccessor} from "mobx-graphlink";
import {GetAccessPolicy} from "../accessPolicies.js";
import {AccessPolicy} from "../accessPolicies/@AccessPolicy.js";
import {APAction, APTable, PermissionSetForType} from "../accessPolicies/@PermissionSet.js";
import {HasAdminPermissions, IsUserCreator, IsUserCreatorOrAdmin, IsUserCreatorOrMod} from "../users/$user.js";
import {Share} from "../shares/@Share.js";
import {User} from "../users/@User.js";
import {Term} from "../terms/@Term.js";
import {Media} from "../media/@Media.js";
import {CommandRun, DMap, GetNode, GetTimeline, MapNodeEdit, NodeL1, NodeLink, NodePhrasing, NodeRating, NodeRevision, NodeTag, Notification, Subscription, Timeline, TimelineStep} from "../../DB.js";

// sync:rs
export const DoPoliciesAllowX = CreateAccessor((actor_id: string | n, policy_targets: string[] | n, action: APAction)=>{
	if (!policy_targets) return false;
	if (policy_targets.length == 0) return false;

	for (const policyTarget of policy_targets) {
		const [policy_id, table] = policyTarget.split(":");
		if (DoesPolicyAllowX(actor_id, policy_id, table as APTable, action)) {
			return true;
		}
	}
	return false;
});

// sync:rs
export const DoesPolicyAllowX = CreateAccessor((actor_id: string | n, policy_id: string | n, table: APTable, action: APAction)=>{
	const policy = GetAccessPolicy(policy_id);
	if (policy == null) return false;

	if (PermissionSetForType.AsBool(policy.permissions[table], action)
		&& PermissionSetForType.AsBool(AccessPolicy.PermissionExtendsForUserAndTable(policy, actor_id, table), action) != false) {
		return true;
	}
	if (PermissionSetForType.AsBool(AccessPolicy.PermissionExtendsForUserAndTable(policy, actor_id, table), action) == true) {
		return true;
	}
	return false;
});

const CanAccessAccessPolicy = (user_id: string|n, policy: AccessPolicy)=>true;
const CanModifyAccessPolicy = (user_id: string|n, policy: AccessPolicy)=>CanAccessAccessPolicy(user_id, policy) && IsUserCreatorOrMod(user_id, policy);
const CanDeleteAccessPolicy = (user_id: string|n, policy: AccessPolicy)=>CanAccessAccessPolicy(user_id, policy) && IsUserCreatorOrMod(user_id, policy);

const CanAccessShare = (user_id: string|n, share: Share)=>true;
const CanModifyShare = (user_id: string|n, share: Share)=>IsUserCreatorOrMod(user_id, share);
const CanDeleteShare = (user_id: string|n, share: Share)=>IsUserCreatorOrMod(user_id, share);

const CanAccessUser = (user_id: string|n, user: User)=>true;
const CanModifyUser = (user_id: string|n, user: User)=>HasAdminPermissions(user_id);
// const CanDeleteUser = (user_id: string|n,user: User)=>HasAdminPermissions(user_id); account deletion not possible atm

const CanAccessMap = (user_id: string|n, map: DMap)=>{
	return IsUserCreatorOrAdmin(user_id, map) || DoesPolicyAllowX(user_id, map.accessPolicy, APTable.maps, APAction.access);
};
const CanModifyMap = (user_id: string|n, map: DMap)=>{
	return CanAccessMap(user_id, map) && (IsUserCreatorOrMod(user_id, map) || DoesPolicyAllowX(user_id, map.accessPolicy, APTable.maps, APAction.modify));
};
const CanDeleteMap = (user_id: string|n, map: DMap)=>{
	return CanAccessMap(user_id, map) && (IsUserCreatorOrMod(user_id, map) || DoesPolicyAllowX(user_id, map.accessPolicy, APTable.maps, APAction.delete));
};

const CanAccessMedia = (user_id: string|n, media: Media|n)=>{
	if (!media) return false;
	return IsUserCreatorOrAdmin(user_id, media) || DoesPolicyAllowX(user_id, media.accessPolicy, APTable.medias, APAction.access);
};
const CanModifyMedia = (user_id: string|n, media: Media|n)=>{
	if (!media) return false;
	return CanAccessMedia(user_id, media) && (IsUserCreatorOrMod(user_id, media) || DoesPolicyAllowX(user_id, media.accessPolicy, APTable.medias, APAction.modify));
};
const CanDeleteMedia = (user_id: string|n, media: Media|n)=>{
	if (!media) return false;
	return CanAccessMedia(user_id, media) && (IsUserCreatorOrMod(user_id, media) || DoesPolicyAllowX(user_id, media.accessPolicy, APTable.medias, APAction.delete));
};

const CanAccessNode = (user_id: string|n, node: NodeL1|n)=>{
	if (!node) return false;
	return IsUserCreatorOrAdmin(user_id, node) || DoesPolicyAllowX(user_id, node.accessPolicy, APTable.nodes, APAction.access);
};
const CanModifyNode = (user_id: string|n, node: NodeL1|n)=>{
	if (!node) return false;
	return CanAccessNode(user_id, node) && (IsUserCreatorOrMod(user_id, node) || DoesPolicyAllowX(user_id, node.accessPolicy, APTable.nodes, APAction.modify));
};
const CanDeleteNode = (user_id: string|n, node: NodeL1|n)=>{
	if (!node) return false;
	return CanAccessNode(user_id, node) && (IsUserCreatorOrMod(user_id, node) || DoesPolicyAllowX(user_id, node.accessPolicy, APTable.nodes, APAction.delete));
};
const CanAddChildNode = (user_id: string|n, node: NodeL1)=>{
	return CanAccessNode(user_id, node) && (IsUserCreatorOrMod(user_id, node) || DoesPolicyAllowX(user_id, node.accessPolicy, APTable.nodes, APAction.addChild));
};
const CanAddPhrasingNode = (user_id: string|n, node: NodeL1)=>{
	return CanAccessNode(user_id, node) && (IsUserCreatorOrMod(user_id, node) || DoesPolicyAllowX(user_id, node.accessPolicy, APTable.nodes, APAction.addPhrasing));
};
const CanVoteNode = (user_id: string|n, node: NodeL1)=>{
	return CanAccessNode(user_id, node) && (IsUserCreatorOrMod(user_id, node) || DoesPolicyAllowX(user_id, node.accessPolicy, APTable.nodes, APAction.vote));
};

const CanAccessTerm = (user_id: string|n, term: Term)=>{
	return IsUserCreatorOrAdmin(user_id, term) || DoesPolicyAllowX(user_id, term.accessPolicy, APTable.terms, APAction.access);
};
const CanModifyTerm = (user_id: string|n, term: Term)=>{
	return CanAccessTerm(user_id, term) && (IsUserCreatorOrMod(user_id, term) || DoesPolicyAllowX(user_id, term.accessPolicy, APTable.terms, APAction.modify));
};
const CanDeleteTerm = (user_id: string|n, term: Term)=>{
	return CanAccessTerm(user_id, term) && (IsUserCreatorOrMod(user_id, term) || DoesPolicyAllowX(user_id, term.accessPolicy, APTable.terms, APAction.delete));
};

const CanAccessTimeline = (user_id: string|n, timeline: Timeline|n)=>{
	if (!timeline) return false;
	return IsUserCreatorOrAdmin(user_id, timeline) || DoesPolicyAllowX(user_id, timeline.accessPolicy, APTable.others, APAction.access);
};
const CanModifyTimeline = (user_id: string|n, timeline: Timeline|n)=>{
	if (!timeline) return false;
	return CanAccessTimeline(user_id, timeline) && (IsUserCreatorOrMod(user_id, timeline) || DoesPolicyAllowX(user_id, timeline.accessPolicy, APTable.others, APAction.modify));
};
const CanDeleteTimeline = (user_id: string|n, timeline: Timeline|n)=>{
	if (!timeline) return false;
	return CanAccessTimeline(user_id, timeline) && (IsUserCreatorOrMod(user_id, timeline) || DoesPolicyAllowX(user_id, timeline.accessPolicy, APTable.others, APAction.delete));
};

const CanAccessNodeLink = (user_id: string|n, nodeLink: NodeLink)=>{
	return IsUserCreatorOrAdmin(user_id, nodeLink) || DoPoliciesAllowX(user_id, nodeLink.c_accessPolicyTargets, APAction.access);
};
const CanModifyNodeLink = (user_id: string|n, nodeLink: NodeLink)=>{
	const child = GetNode(nodeLink.child);
	return IsUserCreatorOrMod(user_id, nodeLink) || (child && CanModifyNode(user_id, child));
};
const CanDeleteNodeLink = (user_id: string|n, nodeLink: NodeLink)=>{
	const child = GetNode(nodeLink.child);
	return IsUserCreatorOrMod(user_id, nodeLink) || (child && CanDeleteNode(user_id, child));
};

const CanAccessNodePhrasing = (user_id: string|n, nodePhrasing: NodePhrasing)=>{
	return IsUserCreatorOrAdmin(user_id, nodePhrasing) || DoPoliciesAllowX(user_id, nodePhrasing.c_accessPolicyTargets, APAction.access);
};
const CanModifyNodePhrasing = (user_id: string|n, nodePhrasing: NodePhrasing)=>{
	return CanAccessNodePhrasing(user_id, nodePhrasing) && IsUserCreatorOrMod(user_id, nodePhrasing);
};
const CanDeleteNodePhrasing = (user_id: string|n, nodePhrasing: NodePhrasing)=>{
	return CanAccessNodePhrasing(user_id, nodePhrasing) && IsUserCreatorOrMod(user_id, nodePhrasing);
};

const CanAccessNodeRating = (user_id: string|n, nodeRating: NodeRating)=>{
	return IsUserCreatorOrAdmin(user_id, nodeRating) || DoPoliciesAllowX(user_id, nodeRating.c_accessPolicyTargets, APAction.access);
};
const CanModifyNodeRating = (user_id: string|n, nodeRating: NodeRating)=>{
	return CanAccessNodeRating(user_id, nodeRating) && IsUserCreator(user_id, nodeRating);
};
const CanDeleteNodeRating = (user_id: string|n, nodeRating: NodeRating)=>{
	return CanAccessNodeRating(user_id, nodeRating) && IsUserCreator(user_id, nodeRating);
};

const CanAccessNodeRevision = (user_id: string|n, nodeRevision: NodeRevision)=>{
	return IsUserCreatorOrAdmin(user_id, nodeRevision) || DoPoliciesAllowX(user_id, nodeRevision.c_accessPolicyTargets, APAction.access);
};
// const CanModifyNodeRevision = (user_id: string|n,nodeRevision: NodeRevision)=> // Revisions cannot be edited
// const CanDeleteNodeRevision = (user_id: string|n,nodeRevision: NodeRevision)=> 

const CanAccessNodeTag = (user_id: string|n, nodeTag: NodeTag)=>{
	return IsUserCreatorOrAdmin(user_id, nodeTag) || DoPoliciesAllowX(user_id, nodeTag.c_accessPolicyTargets, APAction.access);
};
const CanModifyNodeTag = (user_id: string|n, nodeTag: NodeTag)=>{
	return CanAccessNodeTag(user_id, nodeTag) && IsUserCreatorOrMod(user_id, nodeTag);
};
const CanDeleteNodeTag = (user_id: string|n, nodeTag: NodeTag)=>{
	return CanAccessNodeTag(user_id, nodeTag) && IsUserCreatorOrMod(user_id, nodeTag);
};

const CanAccessTimelineStep = (user_id: string|n, timelineStep: TimelineStep)=>{
	return IsUserCreatorOrAdmin(user_id, timelineStep) || DoPoliciesAllowX(user_id, timelineStep.c_accessPolicyTargets, APAction.access);
};
const CanModifyTimelineStep = (user_id: string|n, timelineStep: TimelineStep)=>{
	if (!CanAccessTimelineStep(user_id, timelineStep)) return false;
	if (IsUserCreatorOrMod(user_id, timelineStep)) return true;
	const timeline = GetTimeline(timelineStep.timelineID);
	if (!timeline) return false;
	return DoesPolicyAllowX(user_id, timeline.accessPolicy, APTable.others, APAction.modify);
};
const CanDeleteTimelineStep = (user_id: string|n, timelineStep: TimelineStep)=>{
	if (!CanAccessTimelineStep(user_id, timelineStep)) return false;
	if (IsUserCreatorOrMod(user_id, timelineStep)) return true;
	const timeline = GetTimeline(timelineStep.timelineID);
	if (!timeline) return false;
	return DoesPolicyAllowX(user_id, timeline.accessPolicy, APTable.others, APAction.delete);
};

const CanAccessMapNodeEdit = (user_id: string|n, mapNodeEdit: MapNodeEdit)=>{
	return HasAdminPermissions(user_id) || DoPoliciesAllowX(user_id, mapNodeEdit.c_accessPolicyTargets, APAction.access);
};
const CanAccessUserHidden = (user_id: string|n, user: User)=>{
	return HasAdminPermissions(user_id) || user.id === user_id;
};
const CanModifyUserHidden = (user_id: string|n, user: User)=>CanAccessUserHidden(user_id, user) && user.id === user_id;

const CanAccessCommandRun = (user_id: string|n, commandRun: CommandRun)=>{
	return IsUserCreatorOrAdmin(user_id, {creator: commandRun.actor}) || (commandRun.public_base && DoPoliciesAllowX(user_id, commandRun.c_accessPolicyTargets, APAction.access));
};
const CanAccessSubscription = (user_id: string|n, subscription: Subscription)=>{
	return IsUserCreatorOrAdmin(user_id, {creator: subscription.user});
};
const CanAccessNotification = (user_id: string|n, notification: Notification)=>{
	return IsUserCreatorOrAdmin(user_id, {creator: notification.user});
};
const CanModifyNotification = (user_id: string|n, notification: Notification)=>{
	return CanAccessNotification(user_id, notification) && IsUserCreatorOrAdmin(user_id, {creator: notification.user});
};

export const PERMISSIONS = {
    AccessPolicy: {
        Access: CanAccessAccessPolicy,
        Modify: CanModifyAccessPolicy,
        Delete: CanDeleteAccessPolicy,
    },
    Share: {
        Access: CanAccessShare,
        Modify: CanModifyShare,
        Delete: CanDeleteShare,
    },
    User: {
        Access: CanAccessUser,
        Modify: CanModifyUser,
    	// Delete: CanDeleteUser, // Uncomment if user deletion is implemented in the future
    },
    Map: {
        Access: CanAccessMap,
        Modify: CanModifyMap,
        Delete: CanDeleteMap,
    },
    Media: {
        Access: CanAccessMedia,
        Modify: CanModifyMedia,
        Delete: CanDeleteMedia,
    },
    Node: {
        Access: CanAccessNode,
        Modify: CanModifyNode,
        Delete: CanDeleteNode,
        AddChild: CanAddChildNode,
        AddPhrasing: CanAddPhrasingNode,
        Vote: CanVoteNode,
    },
    Term: {
        Access: CanAccessTerm,
        Modify: CanModifyTerm,
        Delete: CanDeleteTerm,
    },
    Timeline: {
        Access: CanAccessTimeline,
        Modify: CanModifyTimeline,
        Delete: CanDeleteTimeline,
    },
    NodeLink: {
        Access: CanAccessNodeLink,
        Modify: CanModifyNodeLink,
        Delete: CanDeleteNodeLink,
    },
    NodePhrasing: {
        Access: CanAccessNodePhrasing,
        Modify: CanModifyNodePhrasing,
        Delete: CanDeleteNodePhrasing,
    },
    NodeRating: {
        Access: CanAccessNodeRating,
        Modify: CanModifyNodeRating,
        Delete: CanDeleteNodeRating,
    },
    NodeRevision: {
        Access: CanAccessNodeRevision,
    	// Modify: CanModifyNodeRevision, // Uncomment if revision modification is allowed
    	// Delete: CanDeleteNodeRevision, // Uncomment if revision deletion is allowed
    },
    NodeTag: {
        Access: CanAccessNodeTag,
        Modify: CanModifyNodeTag,
        Delete: CanDeleteNodeTag,
    },
    TimelineStep: {
        Access: CanAccessTimelineStep,
        Modify: CanModifyTimelineStep,
        Delete: CanDeleteTimelineStep,
    },
    MapNodeEdit: {
        Access: CanAccessMapNodeEdit,
    },
    UserHidden: {
        Access: CanAccessUserHidden,
        Modify: CanModifyUserHidden,
    },
    CommandRun: {
        Access: CanAccessCommandRun,
    },
    Subscription: {
        Access: CanAccessSubscription,
    },
    Notification: {
        Access: CanAccessNotification,
        Modify: CanModifyNotification,
    },
} as const;

File: Packages/app-server/src/db/_shared/table_permissions.rs
#![cfg_attr(rustfmt, rustfmt_skip)]

use rust_shared::{utils::{auth::jwt_utils_base::UserJWTData, general_::extensions::ToOwnedV}, anyhow::{bail, anyhow}, anyhow::Error};
use tracing::info;

use crate::{db::{terms::Term, access_policies::{get_access_policy}, map_node_edits::MapNodeEdit, user_hiddens::UserHidden, command_runs::CommandRun, node_tags::NodeTag, node_revisions::NodeRevision, node_ratings::NodeRating, node_phrasings::NodePhrasing, node_links::NodeLink, nodes_::_node::Node, maps::Map, medias::Media, feedback_proposals::Proposal, shares::Share, global_data::GlobalData, users::User, access_policies_::{_permission_set::{APAction, APTable}, _access_policy::AccessPolicy}, _shared::access_policy_target::AccessPolicyTarget, nodes::get_node, general::permission_helpers::is_user_admin, feedback_user_infos::UserInfo, timelines::{Timeline, get_timeline}, timeline_steps::TimelineStep}, links::db_live_cache::get_access_policy_cached, utils::db::{accessors::AccessorContext, rls::rls_policies::UsesRLS}};

// Why are the permission-checks for modifying/deleting defined here, rather than in the updateX and deleteX command endpoints?
// Because we want all such logic to use the `can_modify` and `can_delete` macros, so that we can be sure that they always start by checking...
// ...for basic rls-access to the entry, before doing table-specific permission-check logic (since the error-message could easily leak data otherwise).

// empty policies (ie. can always be viewed by anyone) [these functions are not needed in sql version]
// ==========

/*impl CanModify for AccessPolicy {
    async fn can_modify(&self, _ctx: &AccessorContext<'_>, actor: &User) -> Result<bool, Error> {
        is_user_mod_or_creator(actor, &self.creator))
    }
}
impl CanDelete for AccessPolicy {
    async fn can_delete(&self, _ctx: &AccessorContext<'_>, actor: &User) -> Result<bool, Error> {
        is_user_mod_or_creator(actor, &self.creator))
    }
}*/
can_modify!(AccessPolicy, self, actor, { is_user_mod_or_creator(actor, &self.creator) });
can_delete!(AccessPolicy, self, actor, { is_user_mod_or_creator(actor, &self.creator) });

can_modify!(Share, self, actor, { is_user_mod_or_creator(actor, &self.creator) });
can_delete!(Share, self, actor, { is_user_mod_or_creator(actor, &self.creator) });

/*can_modify!(GlobalData, self, actor, { is_user_mod_or_creator(actor, &self.creator });
can_delete!(GlobalData, self, actor, { is_user_mod_or_creator(actor, &self.creator });*/

can_modify!(User, self, actor, { is_user_admin(actor) || actor.id == self.id });
//can_delete!(User, self, actor, { Ok(false }); // account deletion will be possible eventually, but too many complications for now

// likely to be removed at some point
// ----------

can_modify!(Proposal, self, actor, { is_user_mod_or_creator(actor, &self.creator) });
can_delete!(Proposal, self, actor, { is_user_mod_or_creator(actor, &self.creator) });

can_modify!(UserInfo, self, actor, { is_user_admin(actor) || actor.id == self.id });
can_delete!(UserInfo, self, actor, { is_user_admin(actor) || actor.id == self.id });

// simple RLS policies (where to access, it must be that: user is admin, user is creator, or entry's RLS policy allows access)
// ==========

can_modify!(Map, self, ctx, actor, { is_user_mod_or_creator(actor, &self.creator) || does_policy_allow_x(ctx, actor_id(actor), &self.accessPolicy, APTable::maps, APAction::modify).await? });
can_delete!(Map, self, ctx, actor, { is_user_mod_or_creator(actor, &self.creator) || does_policy_allow_x(ctx, actor_id(actor), &self.accessPolicy, APTable::maps, APAction::delete).await? });

can_modify!(Media, self, ctx, actor, { is_user_mod_or_creator(actor, &self.creator) || does_policy_allow_x(ctx, actor_id(actor), &self.accessPolicy, APTable::medias, APAction::modify).await? });
can_delete!(Media, self, ctx, actor, { is_user_mod_or_creator(actor, &self.creator) || does_policy_allow_x(ctx, actor_id(actor), &self.accessPolicy, APTable::medias, APAction::delete).await? });

can_modify!(Node, self, ctx, actor, { is_user_mod_or_creator(actor, &self.creator) || does_policy_allow_x(ctx, actor_id(actor), &self.accessPolicy, APTable::nodes, APAction::modify).await? });
can_delete!(Node, self, ctx, actor, { is_user_mod_or_creator(actor, &self.creator) || does_policy_allow_x(ctx, actor_id(actor), &self.accessPolicy, APTable::nodes, APAction::delete).await? });
impl CanAddChild for Node {
    async fn can_add_child(&self, ctx: &AccessorContext<'_>, actor: &User) -> Result<bool, Error> {
        if !can_access(actor, self) { return Ok(false); }
        Ok(is_user_mod_or_creator(actor, &self.creator) || does_policy_allow_x(ctx, actor_id(actor), &self.accessPolicy, APTable::nodes, APAction::addChild).await?)
    }
}
impl CanAddPhrasing for Node {
    async fn can_add_phrasing(&self, ctx: &AccessorContext<'_>, actor: &User) -> Result<bool, Error> {
        if !can_access(actor, self) { return Ok(false); }
        Ok(is_user_mod_or_creator(actor, &self.creator) || does_policy_allow_x(ctx, actor_id(actor), &self.accessPolicy, APTable::nodes, APAction::addPhrasing).await?)
    }
}
impl CanVote for Node {
    async fn can_vote(&self, ctx: &AccessorContext<'_>, actor: &User) -> Result<bool, Error> {
        if !can_access(actor, self) { return Ok(false); }
        Ok(is_user_creator(actor, &self.creator) || does_policy_allow_x(ctx, actor_id(actor), &self.accessPolicy, APTable::nodes, APAction::vote).await?)
    }
}

can_modify!(Term, self, ctx, actor, { is_user_mod_or_creator(actor, &self.creator) || does_policy_allow_x(ctx, actor_id(actor), &self.accessPolicy, APTable::terms, APAction::modify).await? });
can_delete!(Term, self, ctx, actor, { is_user_mod_or_creator(actor, &self.creator) || does_policy_allow_x(ctx, actor_id(actor), &self.accessPolicy, APTable::terms, APAction::delete).await? });

can_modify!(Timeline, self, ctx, actor, { is_user_mod_or_creator(actor, &self.creator) || does_policy_allow_x(ctx, actor_id(actor), &self.accessPolicy, APTable::others, APAction::modify).await? });
can_delete!(Timeline, self, ctx, actor, { is_user_mod_or_creator(actor, &self.creator) || does_policy_allow_x(ctx, actor_id(actor), &self.accessPolicy, APTable::others, APAction::delete).await? });

// derivative RLS policies (where to access, it must be that: user is admin, user is creator, or all of the associated RLS policies must pass)
// ==========

can_modify!(NodeLink, self, ctx, actor, {
    let child = get_node(ctx, &self.child).await?;
    is_user_mod_or_creator(actor, &self.creator) || child.can_modify(ctx, actor).await?
});
can_delete!(NodeLink, self, ctx, actor, {
    let child = get_node(ctx, &self.child).await?;
    is_user_mod_or_creator(actor, &self.creator) || child.can_delete(ctx, actor).await?
});

can_modify!(NodePhrasing, self, actor, { is_user_mod_or_creator(actor, &self.creator) });
can_delete!(NodePhrasing, self, actor, { is_user_mod_or_creator(actor, &self.creator) });

// only the creator of a rating can edit/delete it
can_modify!(NodeRating, self, actor, { is_user_creator(actor, &self.creator) });
can_delete!(NodeRating, self, actor, { is_user_creator(actor, &self.creator) });

//can_modify!(NodeRevision, self, actor, { is_user_creator(actor, &self.creator) }); // commented; revisions should never be editable, to preserve history
// deletion of non-current node-revisions is allowed -- for dev purposes, or in case sensitive data is accidentally added (in future, will add better handling for this case)
can_delete!(NodeRevision, self, ctx, actor, {
	let node = get_node(&ctx, &self.node).await?;

	let base_text = format!("Cannot delete node-revision #{}, since ", self.id.as_str());
	if node.c_currentRevision == self.id.as_str() {
		bail!("{base_text}it's the current-revision for the node.");
	}
    
    //is_user_admin_or_creator(actor, &self.creator)
    // for now at least, only allow admins to delete node-revisions, to keep history more comprehensive/trustable (if someone accidentally adds sensitive data, they can ask an admin to delete it)
    actor.permissionGroups.admin
});

// only the creator of a node-tag can edit/delete it
can_modify!(NodeTag, self, actor, { is_user_mod_or_creator(actor, &self.creator) });
can_delete!(NodeTag, self, actor, { is_user_mod_or_creator(actor, &self.creator) });

// modify/delete permissions for each timeline-step is the same as that of the timeline it is part of
can_modify!(TimelineStep, self, ctx, actor, {
    is_user_mod_or_creator(actor, &self.creator) || {
        let timeline = get_timeline(ctx, &self.timelineID).await?;
        does_policy_allow_x(ctx, actor_id(actor), &timeline.accessPolicy, APTable::others, APAction::modify).await?
    }
});
can_delete!(TimelineStep, self, ctx, actor, {
    is_user_mod_or_creator(actor, &self.creator) || {
        let timeline = get_timeline(ctx, &self.timelineID).await?;
        does_policy_allow_x(ctx, actor_id(actor), &timeline.accessPolicy, APTable::others, APAction::delete).await?
    }
});

// unique RLS policies
// ==========

/*can_modify!(MapNodeEdit, self, actor, { Ok(false });
can_delete!(MapNodeEdit, self, actor, { Ok(false });*/

// only the given user can edit their own hidden-data
can_modify!(UserHidden, self, actor, { actor.id == self.id });
//can_delete!(UserHidden, self, actor, { Ok(false }); // account deletion will be possible eventually, but too many complications for now

/*can_modify!(CommandRun, self, actor, { Ok(false });
can_delete!(CommandRun, self, actor, { Ok(false });*/

// local helper macros
// ==========

//async fn can_access(&self, ctx: &AccessorContext<'_>, actor: &User) -> Result<bool, Error>;
pub trait CanModify {
    async fn can_modify(&self, ctx: &AccessorContext<'_>, actor: &User) -> Result<bool, Error>;
}
pub trait CanDelete {
    async fn can_delete(&self, ctx: &AccessorContext<'_>, actor: &User) -> Result<bool, Error>;
}
pub trait CanVote {
    async fn can_vote(&self, ctx: &AccessorContext<'_>, actor: &User) -> Result<bool, Error>;
}
pub trait CanAddChild {
    async fn can_add_child(&self, ctx: &AccessorContext<'_>, actor: &User) -> Result<bool, Error>;
}
pub trait CanAddPhrasing {
    async fn can_add_phrasing(&self, ctx: &AccessorContext<'_>, actor: &User) -> Result<bool, Error>;
}

macro_rules! can_modify {
    ($type:ident, $sel:ident, $actor:ident, $body:block) => { can_modify!($type, $sel, _ctx, $actor, $body); };
    ($type:ident, $sel:ident, $ctx:ident, $actor:ident, $body:block) => {
        impl CanModify for $type {
            async fn can_modify(&$sel, $ctx: &AccessorContext<'_>, $actor: &User) -> Result<bool, Error> {
                if !can_access($actor, $sel) { return Ok(false); }
                Ok($body)
            }
        }
    };
}
use can_modify;
macro_rules! can_delete {
    ($type:ident, $sel:ident, $actor:ident, $body:block) => { can_delete!($type, $sel, _ctx, $actor, $body); };
    ($type:ident, $sel:ident, $ctx:ident, $actor:ident, $body:block) => {
        impl CanDelete for $type {
            async fn can_delete(&$sel, $ctx: &AccessorContext<'_>, $actor: &User) -> Result<bool, Error> {
                if !can_access($actor, $sel) { return Ok(false); }
                Ok($body)
            }
        }
    }
}
use can_delete;

// local helpers functions (to match format of rls_policies.rs as much as possible)
// ==========

fn can_access(actor: &User, target: &impl UsesRLS) -> bool {
    target.can_access_cached(Some(actor.id.as_str()))
}
fn actor_id(actor: &User) -> Option<&str> {
    Some(actor.id.as_str())
}

fn is_user_creator(actor: &User, target_creator: &str) -> bool {
    actor.id == target_creator
}
fn is_user_mod_or_creator(actor: &User, target_creator: &str) -> bool {
    if actor.permissionGroups.r#mod { return true; }
    if actor.id == target_creator { return true; }
    return false;
}
fn is_user_admin_or_creator(actor: &User, target_creator: &str) -> bool {
    if actor.permissionGroups.admin { return true; }
    if actor.id == target_creator { return true; }
    return false;
}

// sync:js
pub async fn do_policies_allow_x(ctx: &AccessorContext<'_>, actor_id: Option<&str>, policy_targets: &Vec<AccessPolicyTarget>, action: APAction) -> Result<bool, Error> {
    // The `c_accessPolicyTargets` fields should always have at least one entry in them; if not, something is wrong, so play it safe and reject access.
	// (Most tables enforce non-emptiness of this field with a row constraint, but nodeTags is an exception; its associated nodes may be deleted, leaving it without any targets.)
	// (This line thus serves to prevent "orphaned node-tags" from being visible by non-admins, as well as a general-purpose "second instance" of the non-emptiness check.)
    if policy_targets.is_empty() {
        return Ok(false);
    }
    
    for target in policy_targets {
        if !does_policy_allow_x(ctx, actor_id, &target.policy_id, target.ap_table, action).await? {
            return Ok(false);
        }
    }

    Ok(true)
}
// sync:js
pub async fn does_policy_allow_x(ctx: &AccessorContext<'_>, actor_id: Option<&str>, policy_id: &str, table: APTable, action: APAction) -> Result<bool, Error> {
    let policy = get_access_policy(ctx, policy_id).await?;
    if policy.permissions.for_table(table).as_bool(action)
        && policy.permission_extends_for_user_and_table(actor_id, table).map(|a| a.as_bool(action).clone()) != Some(false) {
        return Ok(true);
    }

    if policy.permission_extends_for_user_and_table(actor_id, table).map(|a| a.as_bool(action)) == Some(true) {
        return Ok(true);
    }

    Ok(false)
}